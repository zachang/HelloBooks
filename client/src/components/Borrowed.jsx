import React from 'react';
import PropTypes from 'react-proptypes';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Pagination } from 'react-materialize';
import { viewUserBorrowAction, returnBookAction } from '../actions/bookAction';
import UserBorrow from './borrow/UserBorrow.jsx';
import { decodeToken } from '../utils/helpers';

/**
 * Borrowed class declaration
 *
 * @class Borrowed
 *
 * @extends {React.Component}
 */
export class Borrowed extends React.Component {
  /**
   * class constructor
   *
   * @param {object} props
   */
  constructor(props) {
    super(props);
    this.state = {
      errors: null,
      books: [],
      pageCount: null,
      limit: 15,
      showToast: false,
    };
    this.returnBook = this.returnBook.bind(this);
  }

  /**
   * @method componentWillMount
   *
   * @return {void} void
   */
  componentWillMount() {
    const userDetails = decodeToken(window.sessionStorage.getItem('token'));
    this.props.viewUserBorrowAction(userDetails.id, this.state.limit, 0);
  }

  /**
   * @method componentWillReceiveProps
   *
   * @param {object} nextProps - nextProps
   *
   * @return {object} nextProps
   */
  componentWillReceiveProps(nextProps) {
    if (nextProps.bookState.success === false) {
      this.setState({ errors: nextProps.bookState.errors });
    }
    if (this.state.pageCount !== nextProps.bookState.pageCount) {
      this.setState({ pageCount: nextProps.bookState.pageCount });
    }
    if (nextProps.bookState.success === true &&
      nextProps.bookState.returned === 'return completed') {
      if (this.state.showToast) {
        Materialize.toast('Book returned successfully,' +
          ' visit library for confirmation!', 4000);
        this.setState({ showToast: false });
      }
    }
    if (nextProps.bookState.fails === 'Error returning book') {
      if (this.state.showToast) {
        Materialize.toast('Return failed, try again!', 5000);
        this.setState({ showToast: false });
      }
    } else if (nextProps.bookState.fails === 'Process failed') {
      if (this.state.showToast) {
        Materialize.toast('Return failed, try again!', 5000);
        this.setState({ showToast: false });
      }
    } else if (nextProps.bookState.fails === 'Book already returned') {
      if (this.state.showToast) {
        Materialize.toast('Book returned already!', 5000);
        this.setState({ showToast: false });
      }
    }
  }

  /**
   * @method componentDidUpdate
   *
   * @return {void} void
   */
  componentDidUpdate() {
    $('.tooltipped').tooltip({ delay: 50 });
  }

  /**
   * Handles book return
   *
   * @method returnBook
   *
   * @return {void}
   *
   * @param {integer} bookId - bookId
   */
  returnBook(bookId) {
    const userDetails = decodeToken(window.sessionStorage.getItem('token'));
    this.props.returnBookAction(userDetails.id, bookId);
    this.setState({ showToast: true });
  }

  /**
   * Renders Borrowed component
   *
   * @return {XML} JSX
   */
  render() {
    const userId = decodeToken(window.sessionStorage.getItem('token'));
    return (
      <div className='row'>
        <div className='section'>
          <h4 style={{ marginTop: '7%' }}>Borrowed Books</h4>
        </div>
        <div className='divider'
          style={{
            marginTop: '-2%',
            marginBottom: '3%'
          }}>
        </div>

        <div className='row'>
          { (this.props.bookState.allBorrows &&
            Object.keys(this.props.bookState.allBorrows).length > 0) ?
            this.props.bookState.allBorrows.map((borrow, i) =>
              <UserBorrow
                key={i}
                borrow={borrow}
                returnBook ={this.returnBook}
              />
            ) :
            <h6 className='no-display'>
              Not borrowed books!
            </h6>
          }

        </div>

        <div className='row'>
          {
            ((this.state.pageCount) ?
              <Pagination
                items={this.state.pageCount}
                onSelect={(page) => {
                  const offset = (page - 1) * this.state.limit;
                  this.props.viewUserBorrowAction(userId.id,
                    this.state.limit, offset);
                }} /> : '')
          }
        </div>
      </div>
    );
  }
}

Borrowed.propTypes = {
  bookState: PropTypes.object.isRequired,
  viewUserBorrowAction: PropTypes.func.isRequired,
  returnBookAction: PropTypes.func.isRequired
};
const mapStateToProps = state => ({
  bookState: state.bookReducer,
});
const mapDispatchToProps = dispatch =>
  bindActionCreators({ viewUserBorrowAction, returnBookAction }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(Borrowed);