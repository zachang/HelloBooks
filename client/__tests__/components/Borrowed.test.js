import React from 'react';
import toJson from 'enzyme-to-json';
import sinon from 'sinon';
import { shallow } from 'enzyme';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import storeConfiguration from '../../src/store/store';
import { Borrowed } from '../../src/components/Borrowed.jsx';

describe('<Borrowed/>', () => {
  // create a spy function for viewUserBorrowAction
  const viewUserBorrowAction = sinon.spy();
  // create a spy function for returnBookAction
  const returnBookAction = sinon.spy();
  // spy on componentWillReceiveProps of Borrowed Page
  sinon.spy(Borrowed.prototype, 'componentWillMount');
  // spy on componentWillReceiveProps of Borrowed Page
  sinon.spy(Borrowed.prototype, 'componentWillReceiveProps');
  // spy on componentDidUpdate of Borrowed Page
  sinon.spy(Borrowed.prototype, 'componentDidUpdate');
  // spy on returnBook of Borrowed Page
  sinon.spy(Borrowed.prototype, 'returnBook');

  const props = {
    bookState: {
      success: true,
      errors: null,
      allBorrows: ['mmm'],
      pageCount: null
    },
    showToast: true,
    viewUserBorrowAction,
    returnBookAction
  };

  const middleware = [thunk];
  const mockStore = configureMockStore(middleware);
  let store;
  let shallowComponent;
  beforeEach(() => {
    store = mockStore(storeConfiguration);
    shallowComponent = shallow(<Borrowed {...props} />);
    shallowComponent.setState({
      pageCount: 1,
      allBorrows: ['mmm'],
      showToast: true
    });
  });

  it('renders <Borrowed /> component', () => {
    const tree = toJson(shallowComponent);
    expect(tree).toMatchSnapshot();
  });

  it('should check that returnBook is called', () => {
    shallowComponent.instance().returnBook(1);
    expect(Borrowed.prototype.returnBook.called).toEqual(true);
  });

  it('calls componentWillReceiveProps when paginating', () => {
    props.pageCount = 1;
    props.bookState.pageCount = 2;
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(Borrowed.prototype.componentWillReceiveProps.called)
      .toEqual(true);
  });

  it(`should check if the 'Pagination' is defined`, () => {
    expect(shallowComponent.find('Pagination').length).toBe(1);
    shallowComponent.find('Pagination').simulate('select');
  });

  it('calls componentWillReceiveProps error when returning books', () => {
    props.bookState.success = false;
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(Borrowed.prototype.componentWillReceiveProps.called).toEqual(true);
  });

  it('calls componentWillReceiveProps when returning books', () => {
    props.bookState.success = true;
    props.bookState.returned = 'return completed';
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(Borrowed.prototype.componentWillReceiveProps.called).toEqual(true);
  });

  it('calls componentWillReceiveProps when returning books', () => {
    props.bookState.success = false;
    props.bookState.fails = 'Error returning book';
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(Borrowed.prototype.componentWillReceiveProps.called).toEqual(true);
  });

  it('calls componentWillReceiveProps when returning books', () => {
    props.bookState.success = false;
    props.bookState.fails = 'Process failed';
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(Borrowed.prototype.componentWillReceiveProps.called).toEqual(true);
  });

  it('calls componentWillReceiveProps when returning books', () => {
    props.bookState.success = false;
    props.bookState.fails = 'Book already returned';
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(Borrowed.prototype.componentWillReceiveProps.called).toEqual(true);
  });
});