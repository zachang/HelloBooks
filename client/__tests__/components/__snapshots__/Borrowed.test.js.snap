// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Borrowed/> renders <Borrowed /> component 1`] = `
<div
  className="row"
>
  <div
    className="section"
  >
    <h4
      style={
        Object {
          "marginTop": "7%",
        }
      }
    >
      Borrowed Books
    </h4>
  </div>
  <div
    className="divider"
    style={
      Object {
        "marginBottom": "3%",
        "marginTop": "-2%",
      }
    }
  />
  <div
    className="row"
  >
    <UserBorrow
      borrow="mmm"
      key="0"
      returnBook={[Function]}
    />
  </div>
  <div
    className="row"
  >
    <Pagination
      activePage={1}
      items={1}
      onSelect={[Function]}
    />
  </div>
</div>
`;
